docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,2624 @@
+
+### Option #3: Auto-rebalance pods in the event of AZ failure
+
+> NOTE: Pre-requisute for this option is to have an ability to detect an outage and recovery of an AZ. 
+
+Only Kube API Server, Gardener Resource Manager and etcd will be setup with multiple replicas spread across zones. Rest of the control plane components will continue to have a single replica. However, in this option etcd cluster members will be rebalanced to ensure that the desired replicas are available at all times.
+
+_Recovering etcd cluster to its full strength_
+
+Affinity rules set on etcd statefulset enforces that there will be at most one etcd member per zone. Two approaches could be taken:
+
+**Variant-#1**
+
+Change the affinity rules and always use `preferredDuringSchedulingIgnoredDuringExecution` for `topologyKey: topology.kubernetes.io/zone`. If all zones are available then it will prefer to distribute the etcd members across zones, each zone having just one replica. In case of zonal failure, kube scheduler will be able to re-schedule this pod in another zone while ensuring that it chooses a node within that zone that does not already have an etcd member running.
+
+_Pros_
+* Simpler to implement as it does not require any change in the affinity rules upon identification of a zonal failure.
+* Etcd cluster runs with full strength as long as there is a single zone where etcd pods can be rescheduled.
+
+_Cons_
+* It is possible that even when there is no zonal failure, more than one etcd member can be provisioned in a single zone. The chances of that happening are slim as typically there is a dedicate worker pool for hosting etcd pods.
+
+**Variant-#2**
+
+Use `requiredDuringSchedulingIgnoredDuringExecution` for `topologyKey: topology.kubernetes.io/zone` during the initial setup to strictly enforce one etcd member per zone. 
+
+If and when a zonal failure is detected then `etcd-druid` should do the following:
+* Remove the PV and PVC for the etcd member in a zone having an outage
+* Change the affinity rules for etcd pods to now use `preferredDuringSchedulingIgnoredDuringExecution` for `topologyKey: topology.kubernetes.io/zone` during the downtime duration of a zone.
+* Delete the etcd pod in the zone which has an outage
+This will force the kube-scheduler to schedule the new pod in another zone.
+
+When it is detected that the zone has now recovered then it should rebalance the etcd members. To achieve that the following `etcd-druid` should do the following:
+* Change the affinity rule to again have `requiredDuringSchedulingIgnoredDuringExecution` for `topologyKey: topology.kubernetes.io/zone`
+* Delete an etcd pod from a zone which has 2 pods running. Subsequently also delete the associated PV and PVC.
+The kube-scheduler will now schedule this pod in the just recovered zone.
+
+Consequence of doing this is that `etcd-druid` which today runs with a single replica, now needs to have a HA setup across zones.
```

Comment:
Hmm... speaking of `etcd-druid`, since it's essential not only in general, but also in the case of DR, shouldn't we run this one too in HA?